File: src/main/java/org/fountainmc/api/command/CommandManager.java
package org.fountainmc.api.command;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.List;

public class CommandManager {

    private final List<CommandHandler> commands;
    private final List<ICommand> dynamicCommands;
    private final List<RegistryHandler> handlers;

    public CommandManager() {
        this.commands = new ArrayList<CommandHandler>();
        this.dynamicCommands = new ArrayList<ICommand>();
        this.handlers = new ArrayList<RegistryHandler>();
    }

    public void registerCommands(Object source) {
        for (Method method : source.getClass().getDeclaredMethods()) {
            if (Modifier.isPublic(method.getModifiers())) {
                if (method.isAnnotationPresent(Command.class)) {
                    Command command = method.getAnnotation(Command.class);
                    CommandHandler cmdHandler = new CommandHandler(command, method, source);
                    commands.add(cmdHandler);
                    for (RegistryHandler handler : handlers) {
                        handler.onRegister(cmdHandler);
                    }
                }
            }
        }
    }

    public void registerHandler(RegistryHandler handler) {
        this.handlers.add(handler);
    }

    public void registerCommand(ICommand iCommand) {
        for (RegistryHandler handler : handlers) {
            handler.onRegister(iCommand);
        }
        dynamicCommands.add(iCommand);
    }


    public void fireCommand(String command, String[] arguments, CommandSender sender) {
        for (CommandHandler handler : commands) {
            if (handler.command.name().equalsIgnoreCase(command)) {
                fireCommand(handler, arguments, sender);
                return;
            }
        }
        for (ICommand dynCmd : dynamicCommands) {
            if (dynCmd.name().equalsIgnoreCase(command)) {
                fireCommand(dynCmd, arguments, sender);
                return;
            }
        }
    }

    public void fireCommand(ICommand command, String[] arguments, CommandSender sender) {
        Class<? extends CommandSender> senderClass = command.allow();
        if (senderClass.isAssignableFrom(sender.getClass())) {
            command.onExecute(senderClass.cast(sender), arguments);
        }
    }

    public void fireCommand(CommandHandler handler, String[] arguments, CommandSender sender) {
        Class<? extends CommandSender> senderClass = handler.command.allow();
        if (senderClass.isAssignableFrom(sender.getClass())) {
            try {
                handler.method.invoke(handler.source, arguments, senderClass.cast(sender));
            } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                e.printStackTrace();
            }
        }
    }

    public List<CommandHandler> getCommands() {
        return commands;
    }

    public class CommandHandler {

        private final Command command;
        private final Method method;
        private final Object source;

        private CommandHandler(Command command, Method method, Object source) {
            this.command = command;
            this.method = method;
            this.source = source;
        }

        public Command getCommand() {
            return command;
        }

        public Method getMethod() {
            return method;
        }

        public Object getSource() {
            return source;
        }
    }

}

File: src/main/kotlin/org/fountainmc/api/command/CommandManager.kt
package org.fountainmc.api.command

import org.fountainmc.api.internal.utils.*
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodType
import java.lang.invoke.WrongMethodTypeException
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.*
import kotlin.reflect.full.isSubclassOf

class CommandManager {

    private val commands: MutableList<CommandHandler>
    private val dynamicCommands: MutableList<AbstractCommand<*>>
    private val handlers: MutableList<RegistryHandler>

    init {
        this.commands = ArrayList<CommandHandler>()
        this.dynamicCommands = ArrayList<AbstractCommand<*>>()
        this.handlers = ArrayList<RegistryHandler>()
    }

    fun registerCommands(source: Any) {
        for (method in source.javaClass.declaredMethods) {
            if (Modifier.isPublic(method.modifiers)) {
                if (method.isAnnotationPresent(Command::class.java)) {
                    val command = method.getAnnotation(Command::class.java)
                    val cmdHandler = CommandHandler(command, method, source)
                    commands.add(cmdHandler)
                    for (handler in handlers) {
                        handler.onRegister(cmdHandler)
                    }
                }
            }
        }
    }

    fun registerHandler(handler: RegistryHandler) {
        this.handlers.add(handler)
    }

    internal fun registerCommand(abstractCommand: AbstractCommand<*>) {
        for (handler in handlers) {
            handler.onRegister(abstractCommand)
        }
        dynamicCommands.add(abstractCommand)
    }

    fun fireCommand(command: String, arguments: Array<String>, sender: CommandSender) {
        for (handler in commands) {
            if (handler.command.name.equals(command, ignoreCase = true)) {
                fireCommand(handler, arguments, sender)
                return
            }
        }
        for (dynCmd in dynamicCommands) {
            if (dynCmd.name().equals(command, ignoreCase = true)) {
                fireCommand(dynCmd, arguments, sender)
                return
            }
        }
    }

    fun <T: CommandSender> fireCommand(command: AbstractCommand<T>, arguments: Array<String>, sender: CommandSender) {
        val senderClass = command.allow()
        if (senderClass.isAssignableFrom(sender.javaClass)) {
            command.onExecute(senderClass.cast(sender), arguments)
        }
    }

    fun fireCommand(handler: CommandHandler, arguments: Array<String>, sender: CommandSender) {
        val allowedSenderClass = handler.command.allow
        if (!allowedSenderClass.isInstance(sender)) {
            sender.sendMessage("[ERROR] Only ${allowedSenderClass.simpleName} may execute this command!")
            return
        }
        try {
            @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN", "USELESS_CAST")
            handler.methodHandle.invokeExact(
                    handler.source as java.lang.Object,
                    arguments as Array<String>,
                    sender
            )
        } catch (e: Throwable) {
            sender.sendMessage("[ERROR] Unexpected ${e.javaClass.simpleName} while executing command!")
            // TODO: Implement proper logging code
            // TODO: Figure out how to determine the plugin to blame
            System.err.println("[ERROR] Unexpected ${e.javaClass.simpleName} while executing command '${handler.command.name}' via ${handler.source.javaClass.typeName}")
            e.printStackTrace()
        }
    }

    fun getCommands(): List<CommandHandler> {
        return commands
    }

    fun getDynamicCommands(): List<AbstractCommand<*>> {
        return dynamicCommands
    }

    class CommandHandler(val command: Command, val method: Method, val source: Any) {
        val methodHandle: MethodHandle
        init {
            require(method.isPublic) { "${method.qualifiedName} isn't publicly accessible" }
            try {
                methodHandle = method.toMethodHandle().asType(HANDLER_METHOD_TYPE)!!
            } catch (e: WrongMethodTypeException) {
                throw IllegalArgumentException("${method.qualifiedName} arguments can't be cast from ${HANDLER_METHOD_TYPE.toMethodDescriptorString()}")
            }
            val senderArgumentType = method.methodType[2].kotlin
            require(command.allow.isSubclassOf(senderArgumentType)) {
                "Allowed sender type ${command.allow.simpleName} must be a subtype of method argument $senderArgumentType"
            }
        }
        companion object {
            val HANDLER_METHOD_TYPE: MethodType = createMethodType(Any::class, Array<String>::class, CommandSender::class)
        }
    }
}
